closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -1070,23 +887,615 @@ boolean isRestrictedProviderAllowed(String providerName) {
+                                                                            )));
+
+            if (debug != null) {
+                // Print information of utilized security profile.
+                rsProperties.listUsedProfile();
+            }
+
+            return rsProperties;
+        }
+
+        /**
+         * Initialize restricted security properties.
+         */
+        private void init(String profileID) {
+            if (debug != null) {
+                debug.println("Initializing restricted security properties for '" + profileID + "'.");
+            }
+
+            String potentialExtendsProfileID = parseProperty(securityProps.getProperty(profileID + ".extends"));
+            if (potentialExtendsProfileID != null) { // If profile extends another profile.
+                if (debug != null) {
+                    debug.println("\t'" + profileID + "' extends '" + potentialExtendsProfileID + "'.");
+                }
+
+                // Check if extended profile exists.
+                String extendsProfileID = null;
+                if (potentialExtendsProfileID.indexOf(".") != potentialExtendsProfileID.lastIndexOf(".")) {
+                    // Extended profile id has at least 2 dots (meaning it's a full profile id).
+                    for (Object keyObject : securityProps.keySet()) {
+                        if (keyObject instanceof String key) {
+                            if (key.startsWith(potentialExtendsProfileID + ".desc") ||
+                                key.startsWith(potentialExtendsProfileID + ".fips") ||
+                                key.startsWith(potentialExtendsProfileID + ".tls") ||
+                                key.startsWith(potentialExtendsProfileID + ".jce") ||
+                                key.startsWith(potentialExtendsProfileID + ".javax") ||
+                                key.startsWith(potentialExtendsProfileID + ".securerandom")
+                            ) {
+                                // If even one security property is found for this profile id, move on.
+                                extendsProfileID = potentialExtendsProfileID;
+                                break;
+                            }
+                        }
+                    }
+                    if (extendsProfileID == null) {
+                        printStackTraceAndExit(potentialExtendsProfileID + " that is supposed to extend '"
+                                + profileID + "' is not present in the java.security file or any appended files.");
+                    }
+                } else {
+                    printStackTraceAndExit(potentialExtendsProfileID + " that is supposed to extend '"
+                            + profileID + "' is not a full profile name.");
+                }
+
+                // Recursively call init() on extended profile.
+                init(extendsProfileID);
```

Comment:
This needs to avoid infinite recursion.

Reply:
We only get here if the profile defines the `.extends` property. So, I'm not sure how this could lead to infinite recursion.

Reply:
I just figured that you mean something like a cyclical extension, where 2 profiles extend each other. So, I added a check to make sure the profile hasn't already been parsed.

Reply:
A cycle might involve one or more profiles; I think the code will handle all those cases.